sympy/integrals/transforms.py

```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
-        ma3 = ma1[g].args[0].collect(t).match(t-b)
-        if ma2 and ma2[a]>0 and ma3 and ma2[a]==ma3[b]:
-            debug('_laplace_apply_rules match:')
-            debug('      f:    %s ( %s, %s, %s )'%(f, ma1, ma2, ma3))
-            debug('      rule: time shift (1.3)')
-            L = _laplace_apply_rules(ma1[g].func(t), t, s, doit=doit, **hints)
-            noconds = hints.get('noconds', False)
-            if not noconds and type(L) is tuple:
-                r, p, c = L
-                return (exp(-ma2[a]*s)*r, p, c)
-            else:
-                return exp(-ma2[a]*s)*L
-    return None
-
+        if ma2 and ma2[a]>0:
+            debug('_laplace_apply_prog_rules match:')
+            debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
+            debug('      rule: time shift (4.1.4)')
+            r, pr, cr = LaplaceTransform(ma1[g].subs(t, t+ma2[a]), t, s).doit(noconds=False)
+            return True, (k*exp(-ma2[a]*s)*r,
+                          Max(p, pr), And(c, cr))
+        if ma2 and ma2[a]<0:
+            debug('_laplace_apply_prog_rules match:')
+            debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
+            debug('      rule: Heaviside factor with negative time shift (4.1.4)')
+            r, pr, cr = LaplaceTransform(ma1[g], t, s).doit(noconds=False)
+            return True, (k*r, Max(p, pr), And(c, cr))
+    return False, (fn, p, c)
 
-def _laplace_rule_exp(f, t, s, doit=True, **hints):
+def _laplace_rule_exp(f, t, s):
     """
-    This internal helper function tries to transform a product containing the
-    `exp` function and returns `None` if it cannot do it.
+    If this function finds a factor ``exp(a*t)``, it applies the
+    frequency-shift rule of the Laplace transform and adjusts the convergence
+    plane accordingly.  For example, if it gets ``(exp(-a*t)*f(t), t, s)``, it
+    will compute ``LaplaceTransform(f(t), t, s+a)``.
     """
-    hints.pop('simplify', True)
-    a = Wild('a', exclude=[t])
+    fn, p, c = _laplace_ct(f)
+    k, f = fn.as_independent(t, as_Add=False)
 
+    a = Wild('a', exclude=[t])
     y = Wild('y')
     z = Wild('z')
     ma1 = f.match(exp(y)*z)
     if ma1:
         ma2 = ma1[y].collect(t).match(a*t)
         if ma2:
-            debug('_laplace_apply_rules match:')
+            debug('_laplace_apply_prog_rules match:')
             debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
-            debug('      rule: multiply with exp (1.5)')
-            L = _laplace_apply_rules(ma1[z], t, s-ma2[a], doit=doit, **hints)
-            noconds = hints.get('noconds', False)
-            if not noconds and type(L) is tuple:
-                r, p, c = L
-                return (r, p+ma2[a], c)
+            debug('      rule: multiply with exp (4.1.5)')
+            r, pr, cr = LaplaceTransform(ma1[z], t, s-ma2[a]).doit(noconds=False)
+            return True, (k*r, Max(p+re(ma2[a]), pr), And(c, cr))
+    return False, (fn, p, c)
+
+def _laplace_rule_delta(f, t, s):
```

Comment:
Can this not be a simple rule?

Reply:
As it is, it can, but later in the development I am also planning to cover mor complicated arguments inside `DiracDelta`, and my instints tell me to keep it separate from the simple rules.